cfspeedtest/cloudflare.py

```diff
@@ -0,0 +1,272 @@
+
+DOWNLOAD_TESTS: TestSpecs = (
+    TestSpec(100_000, 10, "100kB", TestType.Down),
+    TestSpec(1_000_000, 8, "1MB", TestType.Down),
+    TestSpec(10_000_000, 6, "10MB", TestType.Down),
+    TestSpec(25_000_000, 4, "25MB", TestType.Down),
+)
+UPLOAD_TESTS: TestSpecs = (
+    TestSpec(100_000, 8, "100kB", TestType.Up),
+    TestSpec(1_000_000, 6, "1MB", TestType.Up),
+    TestSpec(10_000_000, 4, "10MB", TestType.Up),
+)
+DEFAULT_TESTS: TestSpecs = (
+    TestSpec(1, 20, "latency", TestType.Down),
+    *DOWNLOAD_TESTS,
+    *UPLOAD_TESTS,
+)
+
+
+class TestResult(NamedTuple):
+    """The result of an individual test."""
+
+    value: Any
+    time: float = time.time()
+
+
+class TestTimers(NamedTuple):
+    """A collection of test timer collections, measured in seconds."""
+
+    full: list[float]
+    """The times taken to prepare and perform the requests."""
+    server: list[float]
+    """The times taken to process the requests as reported by the worker."""
+    request: list[float]
+    """The internal client times elapsed to complete the requests."""
+
+    def to_speeds(self, test: TestSpec) -> list[int]:
+        """Compute the test speeds in bits per second from its type and size."""
+        if test.type == TestType.Up:
+            return [int(test.bits / server_time) for server_time in self.server]
+        return [
+            int(test.bits / (full_time - server_time))
+            for full_time, server_time in zip(self.full, self.server)
+        ]
+
+    def to_latencies(self) -> list[float]:
+        """Compute the test latencies in milliseconds."""
+        return [
+            (request_time - server_time) * 1e3
+            for request_time, server_time in zip(self.request, self.server)
+        ]
+
+    @staticmethod
+    def jitter_from(latencies: list[float]) -> float | None:
+        """Compute jitter as average deviation between consecutive latencies."""
+        if len(latencies) < 2:
+            return None
+        return statistics.mean(
+            [
+                abs(latencies[i] - latencies[i - 1])
+                for i in range(1, len(latencies))
+            ]
+        )
+
+
+class TestMetadata(NamedTuple):
+    """The metadata of a test suite."""
+
+    ip: str
+    isp: str
+    location_code: str
+    region: str
+    city: str
+
+
+def _calculate_percentile(data: list[float], percentile: float) -> float:
+    """Find the percentile of a list of values."""
+    data = sorted(data)
+    idx = (len(data) - 1) * percentile
+    rem = idx % 1
+
+    if rem == 0:
+        return data[int(idx)]
+
+    edges = (data[int(idx)], data[int(idx) + 1])
+    return edges[0] + (edges[1] - edges[0]) * rem
+
+
+SuiteResults = dict[str, dict[str, TestResult]]
+
+
+class CloudflareSpeedtest:
+    """Suite of speedtests."""
+
+    def __init__(  # noqa: D417
+        self,
+        results: SuiteResults | None = None,
+        tests: TestSpecs = DEFAULT_TESTS,
+        timeout: tuple[float, float] | float = (3.05, 25),
```

Comment:
Sorry @Nytelife26 ,

I was about to merge but received some TimeoutErrors in my last tests.
It happens for the uploads and I had to change the 3.05 seconds to around 9 seconds to avoid it.

To be safe, we should consider putting this to 25 for connect and get.

Or implement some error handling and retry logic with increased timeout.

What do you think is the best approach?

Reply:
Sorry, I haven't had much free time the past few days. Is this still an issue? The timeout is configurable, but it is always possible for requests to exceed the timeout on slower networks. If you'd like to decide on a sensible lower end bandwidth, below which an error is raised, we can implement that. Other than that, I'm not sure it matters too much.

Reply:
No problem. I figured and decided to merge and just perform this small change myself.

I seem to have the problem hat my connect resolve takes more than 4 seconds sometimes. Something I need to investigate at my Pi-hole. I found that around 9-10 seconds was the sweet spot where I didn't receive the error. Hence, I changed the connect TimeOut to 10 seconds.